src/vocabulary.rs

```diff
@@ -0,0 +1,519 @@
+    /// Turn the vocabulary into datoms, transact them, and on success return the outcome.
+    fn install_vocabulary(&mut self, definition: &Definition) -> Result<VocabularyOutcome> {
+        let (terms, tempids) = definition.description(self)?;
+        self.transact_terms(terms, tempids)?;
+        Ok(VocabularyOutcome::Installed)
+    }
+
+    fn install_attributes_for<'definition>(&mut self, definition: &'definition Definition, attributes: Vec<&'definition (NamespacedKeyword, Attribute)>) -> Result<VocabularyOutcome> {
+        let (terms, tempids) = definition.description_for_attributes(&attributes, self)?;
+        self.transact_terms(terms, tempids)?;
+        Ok(VocabularyOutcome::InstalledMissing)
+    }
+
+    /// Turn the declarative parts of the vocabulary into alterations. Run the 'pre' steps.
+    /// Transact the changes. Run the 'post' steps. Return the result and the new `InProgress`!
+    fn upgrade_vocabulary(&mut self, _definition: &Definition, _from_version: Vocabulary) -> Result<VocabularyOutcome> {
+        unimplemented!();
+        // TODO
+        // Ok(VocabularyOutcome::Installed)
+    }
+}
+
+impl<T> HasVocabularies for T where T: HasSchema + Queryable {
+    fn read_vocabulary_named(&self, name: &NamespacedKeyword) -> Result<Option<Vocabulary>> {
+        if let Some(entid) = self.get_entid(name) {
+            match self.lookup_value_for_attribute(entid, &DB_SCHEMA_VERSION)? {
+                None => Ok(None),
+                Some(TypedValue::Long(version))
+                    if version > 0 && (version < u32::max_value() as i64) => {
+                        let version = version as u32;
+                        let attributes = self.lookup_values_for_attribute(entid, &DB_SCHEMA_ATTRIBUTE)?
+                                             .into_iter()
+                                             .filter_map(|a| {
+                                                 if let TypedValue::Ref(a) = a {
+                                                     self.attribute_for_entid(a)
+                                                         .cloned()
+                                                         .map(|attr| (a, attr))
+                                                 } else {
+                                                     None
+                                                 }
+                                             })
+                                             .collect();
+                        Ok(Some(Vocabulary {
+                            entity: entid.into(),
+                            version: version,
+                            attributes: attributes,
+                        }))
+                    },
+                Some(_) => bail!(ErrorKind::InvalidVocabularyVersion),
+            }
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn read_vocabularies(&self) -> Result<Vocabularies> {
+        // This would be way easier with pull.
```

Comment:
nit: "Datomic's pull syntax" -- not everybody reading this code has full context.